core/arch/riscv64/riscv64.asm

```diff
@@ -118,29 +128,140 @@ GLOBAL_LABEL(dr_call_on_clean_stack:)
+        fsd      f20, 50*ARG_SZ(ARG1)
+        fsd      f21, 51*ARG_SZ(ARG1)
+        fsd      f22, 52*ARG_SZ(ARG1)
+        fsd      f23, 53*ARG_SZ(ARG1)
+        fsd      f24, 54*ARG_SZ(ARG1)
+        fsd      f25, 55*ARG_SZ(ARG1)
+        fsd      f26, 56*ARG_SZ(ARG1)
+        fsd      f27, 57*ARG_SZ(ARG1)
+        fsd      f28, 58*ARG_SZ(ARG1)
+        fsd      f29, 59*ARG_SZ(ARG1)
+        fsd      f30, 60*ARG_SZ(ARG1)
+        fsd      f31, 61*ARG_SZ(ARG1)
+        frcsr    x3
+        sd       x3,  62*ARG_SZ(ARG1)
+        /* No need to save simd registers, at least for now. */
+        ret
+
         DECLARE_EXPORTED_FUNC(dr_app_start)
 GLOBAL_LABEL(dr_app_start:)
-/* FIXME i#3544: Not implemented */
+        /* Save link register for the case that DR is not taking over. */
+        addi     sp, sp, -16
+        sd       ra, 0(sp)
+
+        /* Preverse stack space. */
+        addi     sp, sp, -PRIV_MCONTEXT_SIZE
+        /* Push x3 on to stack to use it as a scratch. */
+        sd       x3, 4*ARG_SZ(sp)
+        /* Compute original sp. */
+        addi     x3, sp, PRIV_MCONTEXT_SIZE+16
+        /* Save link register on to stack. */
+        sd       ra, 1*ARG_SZ(sp)
+        /* Save original sp on to stack. */
+        sd       x3, 2*ARG_SZ(sp)
+        /* Save ra as pc */
+        sd       ra, 32*ARG_SZ(sp)
+        CALLC1(save_priv_mcontext_helper, sp)
+        CALLC1(GLOBAL_REF(dr_app_start_helper), sp)
+        /* If we get here, DR is not taking over. */
+        addi     sp, sp, PRIV_MCONTEXT_SIZE
+        ld       ra, 0(sp)
+        addi     sp, sp, 16
         ret
         END_FUNC(dr_app_start)
 
         DECLARE_EXPORTED_FUNC(dr_app_take_over)
 GLOBAL_LABEL(dr_app_take_over:)
-/* FIXME i#3544: Not implemented */
         j        GLOBAL_REF(dynamorio_app_take_over)
         END_FUNC(dr_app_take_over)
 
         DECLARE_EXPORTED_FUNC(dr_app_running_under_dynamorio)
 GLOBAL_LABEL(dr_app_running_under_dynamorio:)
-/* FIXME i#3544: Not implemented */
+        li       ARG1, 0 /* This instruction is manged by mangle_pre_client. */
         ret
         END_FUNC(dr_app_running_under_dynamorio)
 
 #endif /* DR_APP_EXPORTS */
 
         DECLARE_EXPORTED_FUNC(dynamorio_app_take_over)
 GLOBAL_LABEL(dynamorio_app_take_over:)
-/* FIXME i#3544: Not implemented */
+        /* Save link register for the case that DR is not taking over. */
+        addi     sp, sp, -16
+        sd       ra, 0(sp)
+
+        /* Preverse stack space. */
+        addi     sp, sp, -PRIV_MCONTEXT_SIZE
+        /* Push x3 on to stack to use it as a scratch. */
+        sd       x3, 4*ARG_SZ(sp)
+        /* Compute original sp. */
+        addi     x3, sp, PRIV_MCONTEXT_SIZE+16
+        /* Save link register on to stack. */
+        sd       ra, 1*ARG_SZ(sp)
+        /* Save original sp on to stack. */
+        sd       x3, 2*ARG_SZ(sp)
+        /* Save ra as pc */
+        sd       ra, 32*ARG_SZ(sp)
+        CALLC1(save_priv_mcontext_helper, sp)
```

Comment:
Same concern here: are we clobbering regs already preserved (since `save_priv_mcontext_helper` will overwrite `1*ARG_SZ(ARG1)` which we wrote above at `sd ra, 1*ARG_SZ(sp)`)? Or am I missing something.

Reply:
Fixed in https://github.com/DynamoRIO/dynamorio/pull/6064/commits/ab6cdc5398253654fe9e68904694549b32e42635.